{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3f9a2153",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "import numpy as np\n",
    "import math\n",
    "import sqlite3\n",
    "\n",
    "from pathlib import Path\n",
    "from typing import Union\n",
    "from collections import defaultdict\n",
    "from sqlalchemy import create_engine\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b7d24f1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total electric vehicle stations in the US for fuel type ELEC: 79738\n",
      "The sorted totals have been saved to: data_query/electric_stations_by_state_0106.json\n"
     ]
    }
   ],
   "source": [
    "# see website for information:\n",
    "# https://developer.nrel.gov/docs/transportation/alt-fuel-stations-v1/all/\n",
    "\n",
    "api_key = 'AWMepE53xJxr8Qu8yP3w6J6PLAf3ye6sO7Fsdvlc'\n",
    "base_url = 'https://developer.nrel.gov/api/alt-fuel-stations/v1.json'\n",
    "\n",
    "# List of state codes\n",
    "state_codes = ['AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', 'FL', 'GA', 'HI', 'ID', 'IL', 'IN', 'IA', 'KS', 'KY', 'LA',\n",
    "               'ME', 'MD', 'MA', 'MI', 'MN', 'MS', 'MO', 'MT', 'NE', 'NV', 'NH', 'NJ', 'NM', 'NY', 'NC', 'ND', 'OH', 'OK',\n",
    "               'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY']\n",
    "\n",
    "# Variable to store the total number of electric vehicle stations in the US\n",
    "total_stations_us = 0\n",
    "\n",
    "# Dictionary to store the totals for each state\n",
    "state_totals = {}\n",
    "\n",
    "for state_code in state_codes:\n",
    "    query_params = {\n",
    "        'fuel_type_code': 'ELEC',\n",
    "        'state': state_code,\n",
    "        'country': 'US',\n",
    "        'api_key': api_key\n",
    "    }\n",
    "\n",
    "    # Make the API request\n",
    "    response = requests.get(base_url, params=query_params)\n",
    "\n",
    "    # Parse the JSON response\n",
    "    data = response.json()\n",
    "\n",
    "    # Check if the request was successful (status code 200)\n",
    "    if response.status_code == 200:\n",
    "        # Get the total number of electric vehicle stations for the specified state\n",
    "        total_stations = data.get('total_results', 0)\n",
    "\n",
    "        # Increment the total stations in the US variable\n",
    "        total_stations_us += total_stations\n",
    "\n",
    "        # Store the result in the dictionary\n",
    "        state_totals[state_code] = total_stations\n",
    "\n",
    "    else:\n",
    "        # Print an error message if the request was not successful\n",
    "        print(f'Error: {response.status_code} - {response.text}')\n",
    "\n",
    "# Sort the dictionary by state in decreasing order based on total stations\n",
    "sorted_totals = dict(sorted(state_totals.items(), key=lambda item: item[1], reverse=True))\n",
    "\n",
    "# Create a folder named 'data_query' if it doesn't exist\n",
    "output_folder = 'data_query'\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Save the sorted totals to a JSON file in the 'data_query' folder\n",
    "state_data_file_path = os.path.join(output_folder, 'electric_stations_by_state_0106.json')\n",
    "with open(state_data_file_path, 'w', encoding='utf-8') as state_data_file:\n",
    "    json.dump(sorted_totals, state_data_file, ensure_ascii=False, indent=4)\n",
    "\n",
    "# Print the total number of electric vehicle stations in the US\n",
    "print(f'Total electric vehicle stations in the US for fuel type ELEC: {total_stations_us}')\n",
    "\n",
    "print(f'The sorted totals have been saved to: {state_data_file_path}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "86d9ed91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total results by state:\n",
      "state\n",
      "CA    19351\n",
      "NY     4391\n",
      "TX     4170\n",
      "FL     4055\n",
      "MA     3201\n",
      "WA     2557\n",
      "CO     2511\n",
      "GA     2298\n",
      "IL     2291\n",
      "PA     2185\n",
      "OH     2039\n",
      "NC     2024\n",
      "MD     2009\n",
      "MI     1917\n",
      "VA     1870\n",
      "MN     1565\n",
      "MO     1543\n",
      "AZ     1463\n",
      "NJ     1438\n",
      "OR     1436\n",
      "TN     1139\n",
      "IA     1099\n",
      "UT     1025\n",
      "WI     1011\n",
      "IN      947\n",
      "CT      932\n",
      "SC      753\n",
      "KS      724\n",
      "OK      707\n",
      "NV      685\n",
      "AL      581\n",
      "ME      513\n",
      "AR      500\n",
      "DC      442\n",
      "KY      441\n",
      "HI      434\n",
      "NE      432\n",
      "VT      414\n",
      "NM      385\n",
      "LA      372\n",
      "RI      350\n",
      "NH      295\n",
      "MS      257\n",
      "ID      248\n",
      "DE      248\n",
      "WV      214\n",
      "SD      198\n",
      "MT      158\n",
      "ND      154\n",
      "WY      142\n",
      "AK       66\n",
      "PR       21\n",
      "KA        1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "api_key = 'AWMepE53xJxr8Qu8yP3w6J6PLAf3ye6sO7Fsdvlc'\n",
    "base_url = 'https://developer.nrel.gov/api/alt-fuel-stations/v1.json'\n",
    "\n",
    "# Set up parameters for the API request\n",
    "params = {\n",
    "    'fuel_type_code': 'ELEC',\n",
    "    'country': 'US',\n",
    "    'restricted_access': 'false',\n",
    "    'api_key': api_key\n",
    "}\n",
    "\n",
    "# Make the API request and convert the response to a DataFrame\n",
    "response = requests.get(base_url, params=params)\n",
    "data = response.json()\n",
    "\n",
    "if response.status_code == 200:\n",
    "    # Extract and normalize the data\n",
    "    fuel_stations_df = pd.json_normalize(data['fuel_stations'])\n",
    "\n",
    "    # Keep only the specified fields in the DataFrame\n",
    "    selected_fields = ['id', 'station_name', 'street_address', 'city', 'state', 'zip',\n",
    "                       'latitude', 'longitude', 'access_code', 'ev_network', 'ev_connector_types', 'status_code',\n",
    "                       'ev_pricing', 'geocode_status']\n",
    "    fuel_stations_df = fuel_stations_df[selected_fields]\n",
    "\n",
    "    # Get the total count of each unique value in the 'state' field\n",
    "    total_results_by_state = fuel_stations_df['state'].value_counts()\n",
    "\n",
    "    print(f'Total results by state:\\n{total_results_by_state}')\n",
    "else:\n",
    "    print(f'Error: {response.status_code} - {response.text}')\n",
    "    exit()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "102d31b7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The fuel station data for the state of California has been saved to: data_query/fuel_stations_selected_by_CA_0106.json\n"
     ]
    }
   ],
   "source": [
    "# API Pull for only California\n",
    "api_key = 'AWMepE53xJxr8Qu8yP3w6J6PLAf3ye6sO7Fsdvlc'\n",
    "base_url = 'https://developer.nrel.gov/api/alt-fuel-stations/v1.json'\n",
    "\n",
    "# Selected fields to retrieve\n",
    "selected_fields = ['id', 'station_name', 'street_address', 'city', 'state', 'zip',\n",
    "                   'latitude', 'longitude', 'access_code', 'ev_network', 'ev_connector_types', 'status_code',\n",
    "                   'ev_pricing', 'geocode_status']\n",
    "\n",
    "# List of states to filter (Only 'CA' in this case)\n",
    "state_codes = ['CA']\n",
    "\n",
    "# Set up parameters for the API request\n",
    "params = {\n",
    "    'fuel_type_code': 'ELEC',\n",
    "    'country': 'US',\n",
    "    'restricted_access': 'false',\n",
    "    'api_key': api_key\n",
    "}\n",
    "\n",
    "# List to store detailed information for stations meeting the criteria\n",
    "station_data_by_state = {}\n",
    "\n",
    "# Function to make API requests and handle errors\n",
    "def make_api_request(url, params):\n",
    "    try:\n",
    "        response = requests.get(url, params=params)\n",
    "        response.raise_for_status()\n",
    "        return response.json() if response.status_code == 200 else None\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        print(f'Error making the API request: {e}')\n",
    "        return None\n",
    "\n",
    "# Make API requests for each state and filter based on state codes\n",
    "for state_code in state_codes:\n",
    "    params['state'] = state_code  # Include state code in the parameters\n",
    "\n",
    "    # Make the API request\n",
    "    data = make_api_request(base_url, params)\n",
    "    \n",
    "    if data is not None:\n",
    "        # Convert the response to a DataFrame with selected fields\n",
    "        stations_df = pd.json_normalize(data['fuel_stations'], sep='_')[selected_fields].fillna('N/A')\n",
    "        station_data_by_state[state_code] = stations_df.to_dict(orient='records')\n",
    "\n",
    "# Create a folder named 'data_query' if it doesn't exist\n",
    "output_folder = 'data_query'\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "# Save the station data by state to a JSON file in the 'data_query' folder\n",
    "output_file_path = os.path.join(output_folder, 'fuel_stations_selected_by_CA_0106.json')\n",
    "with open(output_file_path, 'w', encoding='utf-8') as output_file:\n",
    "    json.dump(station_data_by_state, output_file, ensure_ascii=False, indent=4)\n",
    "\n",
    "print(f'The fuel station data for the state of California has been saved to: {output_file_path}')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e41b563f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in JSON data from the file\n",
    "json_data_path = Path(\"data_query/fuel_stations_selected_by_CA_0106.json\")\n",
    "\n",
    "if json_data_path.exists():\n",
    "    with json_data_path.open(\"r\", encoding=\"utf8\") as f:\n",
    "        fuel_station_data = json.load(f)\n",
    "\n",
    "# Check if the necessary data is present in the loaded JSON\n",
    "if 'CA' in fuel_station_data:\n",
    "    # Extract data for California from the loaded JSON\n",
    "    ca_station_data = fuel_station_data['CA']\n",
    "\n",
    "    # Create DataFrame\n",
    "    fuel_station_df = pd.DataFrame(ca_station_data)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "72818372",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id                      int64\n",
       "station_name           object\n",
       "street_address         object\n",
       "city                   object\n",
       "state                  object\n",
       "zip                    object\n",
       "latitude              float64\n",
       "longitude             float64\n",
       "access_code            object\n",
       "ev_network             object\n",
       "ev_connector_types     object\n",
       "status_code            object\n",
       "ev_pricing             object\n",
       "geocode_status         object\n",
       "dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fuel_station_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d38faaf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    id                        station_name         street_address  \\\n",
      "0  792  PG&E - Grass Valley Service Center     788 Taylorville Rd   \n",
      "1  798    PG&E - Santa Cruz Service Center            615 7th Ave   \n",
      "2  801   PG&E - Bakersfield Service Center          4101 Wible Rd   \n",
      "3  806        PG&E - Fresno Service Center  3754 E California Ave   \n",
      "4  809       PG&E - Salinas Service Center        390 E Alisal St   \n",
      "\n",
      "           city state      zip   latitude   longitude access_code ev_network  \\\n",
      "0  Grass Valley    CA  95949.0  39.200620 -121.062126      public        N/A   \n",
      "1    Santa Cruz    CA  95062.0  36.969267 -121.999295      public        N/A   \n",
      "2   Bakersfield    CA  93313.0  35.317617 -119.039048      public        N/A   \n",
      "3        Fresno    CA  93725.0  36.721360 -119.761165      public        N/A   \n",
      "4       Salinas    CA  93901.0  36.672170 -121.641106      public        N/A   \n",
      "\n",
      "  ev_connector_types status_code ev_pricing geocode_status  \n",
      "0                N/A           E        N/A          200-9  \n",
      "1                N/A           E        N/A          200-9  \n",
      "2                N/A           E        N/A          200-8  \n",
      "3                N/A           E        N/A          200-8  \n",
      "4                N/A           E        N/A          200-8  \n"
     ]
    }
   ],
   "source": [
    "# Data cleansing operations\n",
    "conversion_mapping = {\n",
    "    \"station_name\": \"string\",\n",
    "    \"street_address\": \"string\",\n",
    "    \"state\": \"string\",\n",
    "    \"city\": \"string\",\n",
    "    \"zip\": \"Int64\",\n",
    "    \"latitude\": \"float64\",\n",
    "    \"longitude\": \"float64\",\n",
    "    \"access_code\": \"string\",\n",
    "    \"ev_network\": \"string\",\n",
    "    \"ev_connector_types\": \"string\",\n",
    "    \"status_code\": \"string\",\n",
    "    \"ev_pricing\": \"string\",\n",
    "    \"geocode_status\": \"string\",\n",
    "\n",
    "}\n",
    "\n",
    "for column, dtype in conversion_mapping.items():\n",
    "    try:\n",
    "        if column == 'zip':\n",
    "            fuel_station_df[column] = pd.to_numeric(fuel_station_df[column], errors='coerce', downcast='integer')\n",
    "        elif column.startswith('ev_') and dtype == 'Int64':\n",
    "            # Convert only if the value is not 'N/A'\n",
    "            fuel_station_df[column] = pd.to_numeric(fuel_station_df[column].replace('N/A', pd.NA), errors='coerce', downcast='integer')\n",
    "        else:\n",
    "            fuel_station_df[column] = fuel_station_df[column].astype(dtype)\n",
    "    except (ValueError, pd.errors.OutOfBoundsDatetime) as e:\n",
    "        print(f\"Error converting column {column} to {dtype}: {e}\")\n",
    "        # Handle the exception here, for example, assign a default value\n",
    "        fuel_station_df[column] = None  # or any default value you prefer\n",
    "\n",
    "# Display the first few rows of the DataFrame\n",
    "print(fuel_station_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0067ecd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id                             int64\n",
       "station_name          string[python]\n",
       "street_address        string[python]\n",
       "city                  string[python]\n",
       "state                 string[python]\n",
       "zip                          float64\n",
       "latitude                     float64\n",
       "longitude                    float64\n",
       "access_code           string[python]\n",
       "ev_network            string[python]\n",
       "ev_connector_types    string[python]\n",
       "status_code           string[python]\n",
       "ev_pricing            string[python]\n",
       "geocode_status        string[python]\n",
       "dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fuel_station_df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "05f6c65b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    id                        station_name         street_address  \\\n",
      "0  792  PG&E - Grass Valley Service Center     788 Taylorville Rd   \n",
      "1  798    PG&E - Santa Cruz Service Center            615 7th Ave   \n",
      "2  801   PG&E - Bakersfield Service Center          4101 Wible Rd   \n",
      "3  806        PG&E - Fresno Service Center  3754 E California Ave   \n",
      "4  809       PG&E - Salinas Service Center        390 E Alisal St   \n",
      "\n",
      "           city state      zip   latitude   longitude access_code ev_network  \\\n",
      "0  Grass Valley    CA  95949.0  39.200620 -121.062126      public        N/A   \n",
      "1    Santa Cruz    CA  95062.0  36.969267 -121.999295      public        N/A   \n",
      "2   Bakersfield    CA  93313.0  35.317617 -119.039048      public        N/A   \n",
      "3        Fresno    CA  93725.0  36.721360 -119.761165      public        N/A   \n",
      "4       Salinas    CA  93901.0  36.672170 -121.641106      public        N/A   \n",
      "\n",
      "  ev_connector_types status_code ev_pricing geocode_status  \n",
      "0                N/A           E        N/A          200-9  \n",
      "1                N/A           E        N/A          200-9  \n",
      "2                N/A           E        N/A          200-8  \n",
      "3                N/A           E        N/A          200-8  \n",
      "4                N/A           E        N/A          200-8  \n"
     ]
    }
   ],
   "source": [
    " # Replace empty strings with None\n",
    "fuel_station_df = fuel_station_df.replace(\"\", None)\n",
    "\n",
    "# Set default values for ID columns if they are null/NaN/None\n",
    "id_columns = [\"id\", \"station_name\", \"street_address\", \"city\", \"state\", \"zip\",\n",
    "              \"latitude\", \"longitude\", \"access_code\", \"ev_network\",\n",
    "              \"ev_connector_types\", \"status_code\", \"ev_pricing\",\n",
    "              \"geocode_status\"]\n",
    "fuel_station_df[id_columns] = fuel_station_df[id_columns].fillna(0)\n",
    "\n",
    "# Replace specific values if needed\n",
    "fuel_station_df[\"ev_pricing\"] = fuel_station_df[\"ev_pricing\"].replace(\"FREE\", \"Free\")\n",
    "\n",
    "# Drop duplicate rows \n",
    "fuel_station_df = fuel_station_df.drop_duplicates()\n",
    "\n",
    "# Display the first few rows of the DataFrame\n",
    "print(fuel_station_df.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f42a5c06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The SQLite database has been created, and the DataFrame has been stored in the 'fuel_stations' table.\n"
     ]
    }
   ],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import json\n",
    "\n",
    "# Read in JSON data from the file\n",
    "json_data_path = Path(\"data_query/fuel_stations_selected_by_CA_0106.json\")\n",
    "\n",
    "if json_data_path.exists():\n",
    "    with json_data_path.open(\"r\", encoding=\"utf8\") as f:\n",
    "        fuel_station_data = json.load(f)\n",
    "\n",
    "    # Check if the necessary data is present in the loaded JSON\n",
    "    if 'CA' in fuel_station_data:\n",
    "        # Extract data for California from the loaded JSON\n",
    "        ca_station_data = fuel_station_data['CA']\n",
    "\n",
    "        # Create DataFrame\n",
    "        fuel_station_df = pd.DataFrame(ca_station_data)\n",
    "\n",
    "        # Clean the DataFrame\n",
    "        fuel_station_df = fuel_station_df.replace(\"\", None)\n",
    "        id_columns = [\"id\", \"station_name\", \"street_address\", \"city\", \"state\", \"zip\",\n",
    "                      \"latitude\", \"longitude\", \"access_code\", \"ev_network\",\n",
    "                      \"ev_connector_types\", \"status_code\", \"ev_pricing\",\n",
    "                      \"geocode_status\"]\n",
    "        fuel_station_df[id_columns] = fuel_station_df[id_columns].fillna(0)\n",
    "        fuel_station_df[\"ev_pricing\"] = fuel_station_df[\"ev_pricing\"].replace(\"FREE\", \"Free\")\n",
    "\n",
    "        # Filter DataFrame to columns with lists\n",
    "        columns_with_lists = fuel_station_df.select_dtypes(include='object').columns[\n",
    "            fuel_station_df.select_dtypes(include='object').applymap(lambda x: isinstance(x, list)).any()\n",
    "        ]\n",
    "        fuel_station_df = fuel_station_df.drop(columns=columns_with_lists)\n",
    "\n",
    "        # Drop duplicate rows\n",
    "        fuel_station_df = fuel_station_df.drop_duplicates()\n",
    "\n",
    "        # Define the path to your SQLite database file\n",
    "        database_path = Path(\"fuel_stations.db\")\n",
    "\n",
    "        # Connect to the SQLite database\n",
    "        conn = sqlite3.connect(database_path)\n",
    "\n",
    "        # Create a cursor object to execute SQL queries\n",
    "        cursor = conn.cursor()\n",
    "\n",
    "        # Define the SQL statement to create a table\n",
    "        create_table_query = \"\"\"\n",
    "        CREATE TABLE IF NOT EXISTS fuel_stations (\n",
    "            id INTEGER PRIMARY KEY,\n",
    "            station_name TEXT,\n",
    "            street_address TEXT,\n",
    "            city TEXT,\n",
    "            state TEXT,\n",
    "            zip INTEGER,\n",
    "            latitude REAL,\n",
    "            longitude REAL,\n",
    "            access_code TEXT,\n",
    "            ev_network TEXT,\n",
    "            ev_connector_types TEXT,\n",
    "            status_code TEXT,\n",
    "            ev_pricing TEXT,\n",
    "            geocode_status TEXT\n",
    "        );\n",
    "        \"\"\"\n",
    "\n",
    "        # Execute the SQL query to create the table\n",
    "        cursor.execute(create_table_query)\n",
    "\n",
    "        # Replace 'N/A' with None for SQLite compatibility\n",
    "        fuel_station_df_cleaned = fuel_station_df.replace({'N/A': None, pd.NA: None})\n",
    "\n",
    "        # Explicitly set data types for problematic columns\n",
    "        numeric_columns = ['zip', 'latitude', 'longitude']\n",
    "        for column in numeric_columns:\n",
    "            fuel_station_df_cleaned[column] = pd.to_numeric(fuel_station_df_cleaned[column], errors='coerce', downcast='integer')\n",
    "\n",
    "        # Convert DataFrame to SQLite table without specifying data types\n",
    "        fuel_station_df_cleaned.to_sql(\"fuel_stations\", conn, if_exists=\"replace\", index=False)\n",
    "\n",
    "        # Commit the changes and close the connection\n",
    "        conn.commit()\n",
    "        conn.close()\n",
    "\n",
    "        print(f\"The SQLite database has been created, and the DataFrame has been stored in the 'fuel_stations' table.\")\n",
    "    else:\n",
    "        print(f\"No data available for California in the loaded JSON.\")\n",
    "else:\n",
    "    print(f\"JSON data file not found.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b802a3fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The DataFrame has been saved to: data_query/fuel_stations_ca.csv\n"
     ]
    }
   ],
   "source": [
    "# Read in JSON data from the file\n",
    "json_data_path = Path(\"data_query/fuel_stations_selected_by_CA_0106.json\")\n",
    "\n",
    "if json_data_path.exists():\n",
    "    with json_data_path.open(\"r\", encoding=\"utf8\") as f:\n",
    "        fuel_station_data = json.load(f)\n",
    "\n",
    "    # Check if the necessary data is present in the loaded JSON\n",
    "    if 'CA' in fuel_station_data:\n",
    "        # Extract data for California from the loaded JSON\n",
    "        ca_station_data = fuel_station_data['CA']\n",
    "\n",
    "        # Create DataFrame\n",
    "        fuel_station_df = pd.DataFrame(ca_station_data)\n",
    "\n",
    "        # Save DataFrame to CSV file\n",
    "        csv_file_path = \"data_query/fuel_stations_ca.csv\"\n",
    "        fuel_station_df.to_csv(csv_file_path, index=False)\n",
    "\n",
    "        print(f\"The DataFrame has been saved to: {csv_file_path}\")\n",
    "else:\n",
    "    print(f\"No data available for California in the loaded JSON.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "72370a6f",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unindent does not match any outer indentation level (<tokenize>, line 49)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m<tokenize>:49\u001b[0;36m\u001b[0m\n\u001b[0;31m    else:\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unindent does not match any outer indentation level\n"
     ]
    }
   ],
   "source": [
    " # Define the path to your SQLite database file\n",
    "        database_path = Path(\"fuel_stations.db\")\n",
    "\n",
    "        # Connect to the SQLite database\n",
    "        conn = sqlite3.connect(database_path)\n",
    "\n",
    "        # Create a cursor object to execute SQL queries\n",
    "        cursor = conn.cursor()\n",
    "\n",
    "        # Define the SQL statement to create a table\n",
    "        create_table_query = \"\"\"\n",
    "        CREATE TABLE IF NOT EXISTS fuel_stations (\n",
    "            id INTEGER PRIMARY KEY,\n",
    "            station_name TEXT,\n",
    "            street_address TEXT,\n",
    "            city TEXT,\n",
    "            state TEXT,\n",
    "            zip INTEGER,\n",
    "            latitude REAL,\n",
    "            longitude REAL,\n",
    "            access_code TEXT,\n",
    "            ev_network TEXT,\n",
    "            ev_connector_types TEXT,\n",
    "            status_code TEXT,\n",
    "            ev_pricing TEXT,\n",
    "            geocode_status TEXT\n",
    "        );\n",
    "        \"\"\"\n",
    "\n",
    "        # Execute the SQL query to create the table\n",
    "        cursor.execute(create_table_query)\n",
    "\n",
    "        # Replace 'N/A' with None for SQLite compatibility\n",
    "        fuel_station_df_cleaned = fuel_station_df.replace({'N/A': None, pd.NA: None})\n",
    "\n",
    "        # Explicitly set data types for problematic columns\n",
    "        numeric_columns = ['zip', 'latitude', 'longitude']\n",
    "        for column in numeric_columns:\n",
    "            fuel_station_df_cleaned[column] = pd.to_numeric(fuel_station_df_cleaned[column], errors='coerce', downcast='integer')\n",
    "\n",
    "        # Convert DataFrame to SQLite table without specifying data types\n",
    "        fuel_station_df_cleaned.to_sql(\"fuel_stations\", conn, if_exists=\"replace\", index=False)\n",
    "\n",
    "        # Commit the changes and close the connection\n",
    "        conn.commit()\n",
    "        conn.close()\n",
    "\n",
    "        print(f\"The SQLite database has been created, and the DataFrame has been stored in the 'fuel_stations' table.\")\n",
    "    else:\n",
    "        print(f\"No data available for California in the loaded JSON.\")\n",
    "    else:\n",
    "    print(f\"JSON data file not found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "16e77dfd",
   "metadata": {},
   "outputs": [
    {
     "ename": "OperationalError",
     "evalue": "(sqlite3.OperationalError) unable to open database file\n(Background on this error at: https://sqlalche.me/e/14/e3q8)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3280\u001b[0m, in \u001b[0;36mEngine._wrap_pool_connect\u001b[0;34m(self, fn, connection)\u001b[0m\n\u001b[1;32m   3279\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3280\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3281\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m dialect\u001b[38;5;241m.\u001b[39mdbapi\u001b[38;5;241m.\u001b[39mError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:310\u001b[0m, in \u001b[0;36mPool.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    303\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Return a DBAPI connection from the pool.\u001b[39;00m\n\u001b[1;32m    304\u001b[0m \n\u001b[1;32m    305\u001b[0m \u001b[38;5;124;03mThe connection is instrumented such that when its\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    308\u001b[0m \n\u001b[1;32m    309\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 310\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_ConnectionFairy\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_checkout\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:868\u001b[0m, in \u001b[0;36m_ConnectionFairy._checkout\u001b[0;34m(cls, pool, threadconns, fairy)\u001b[0m\n\u001b[1;32m    867\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fairy:\n\u001b[0;32m--> 868\u001b[0m     fairy \u001b[38;5;241m=\u001b[39m \u001b[43m_ConnectionRecord\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcheckout\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpool\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    870\u001b[0m     fairy\u001b[38;5;241m.\u001b[39m_pool \u001b[38;5;241m=\u001b[39m pool\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:476\u001b[0m, in \u001b[0;36m_ConnectionRecord.checkout\u001b[0;34m(cls, pool)\u001b[0m\n\u001b[1;32m    474\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[1;32m    475\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcheckout\u001b[39m(\u001b[38;5;28mcls\u001b[39m, pool):\n\u001b[0;32m--> 476\u001b[0m     rec \u001b[38;5;241m=\u001b[39m \u001b[43mpool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_do_get\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    477\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/impl.py:256\u001b[0m, in \u001b[0;36mNullPool._do_get\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    255\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_do_get\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 256\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create_connection\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:256\u001b[0m, in \u001b[0;36mPool._create_connection\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    254\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Called by subclasses to create a new ConnectionRecord.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 256\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_ConnectionRecord\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:371\u001b[0m, in \u001b[0;36m_ConnectionRecord.__init__\u001b[0;34m(self, pool, connect)\u001b[0m\n\u001b[1;32m    370\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m connect:\n\u001b[0;32m--> 371\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m__connect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    372\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfinalize_callback \u001b[38;5;241m=\u001b[39m deque()\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:665\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    664\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 665\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m util\u001b[38;5;241m.\u001b[39msafe_reraise():\n\u001b[1;32m    666\u001b[0m         pool\u001b[38;5;241m.\u001b[39mlogger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError on connect(): \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, e)\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py:70\u001b[0m, in \u001b[0;36msafe_reraise.__exit__\u001b[0;34m(self, type_, value, traceback)\u001b[0m\n\u001b[1;32m     69\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwarn_only:\n\u001b[0;32m---> 70\u001b[0m         \u001b[43mcompat\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mraise_\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     71\u001b[0m \u001b[43m            \u001b[49m\u001b[43mexc_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     72\u001b[0m \u001b[43m            \u001b[49m\u001b[43mwith_traceback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexc_tb\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     73\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/compat.py:208\u001b[0m, in \u001b[0;36mraise_\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    207\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 208\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exception\n\u001b[1;32m    209\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m     \u001b[38;5;66;03m# credit to\u001b[39;00m\n\u001b[1;32m    211\u001b[0m     \u001b[38;5;66;03m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[1;32m    212\u001b[0m     \u001b[38;5;66;03m# as the __traceback__ object creates a cycle\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:661\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    660\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstarttime \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[0;32m--> 661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdbapi_connection \u001b[38;5;241m=\u001b[39m connection \u001b[38;5;241m=\u001b[39m \u001b[43mpool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_invoke_creator\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    662\u001b[0m pool\u001b[38;5;241m.\u001b[39mlogger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCreated new connection \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, connection)\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/create.py:590\u001b[0m, in \u001b[0;36mcreate_engine.<locals>.connect\u001b[0;34m(connection_record)\u001b[0m\n\u001b[1;32m    589\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m connection\n\u001b[0;32m--> 590\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdialect\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcparams\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:597\u001b[0m, in \u001b[0;36mDefaultDialect.connect\u001b[0;34m(self, *cargs, **cparams)\u001b[0m\n\u001b[1;32m    595\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39mcargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mcparams):\n\u001b[1;32m    596\u001b[0m     \u001b[38;5;66;03m# inherits the docstring from interfaces.Dialect.connect\u001b[39;00m\n\u001b[0;32m--> 597\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdbapi\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcparams\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mOperationalError\u001b[0m: unable to open database file",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 62\u001b[0m\n\u001b[1;32m     59\u001b[0m database_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdata/ca_data.sqlite\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     61\u001b[0m manager \u001b[38;5;241m=\u001b[39m SQLiteManager(database_path)\n\u001b[0;32m---> 62\u001b[0m \u001b[43mmanager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_database\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     63\u001b[0m manager\u001b[38;5;241m.\u001b[39mimport_csv_to_table(csv_file_path)\n\u001b[1;32m     65\u001b[0m \u001b[38;5;66;03m# Example query\u001b[39;00m\n\u001b[1;32m     66\u001b[0m \u001b[38;5;66;03m# results = manager.session.query(ReducedData).filter_by(city='YourCity').all()\u001b[39;00m\n\u001b[1;32m     67\u001b[0m \u001b[38;5;66;03m# for result in results:\u001b[39;00m\n\u001b[1;32m     68\u001b[0m \u001b[38;5;66;03m#     print(result.station_name, result.street_address)\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[23], line 35\u001b[0m, in \u001b[0;36mSQLiteManager.create_database\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcreate_database\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m---> 35\u001b[0m     \u001b[43mBase\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_all\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/sql/schema.py:4916\u001b[0m, in \u001b[0;36mMetaData.create_all\u001b[0;34m(self, bind, tables, checkfirst)\u001b[0m\n\u001b[1;32m   4914\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m bind \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   4915\u001b[0m     bind \u001b[38;5;241m=\u001b[39m _bind_or_error(\u001b[38;5;28mself\u001b[39m)\n\u001b[0;32m-> 4916\u001b[0m \u001b[43mbind\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run_ddl_visitor\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   4917\u001b[0m \u001b[43m    \u001b[49m\u001b[43mddl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mSchemaGenerator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheckfirst\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcheckfirst\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtables\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtables\u001b[49m\n\u001b[1;32m   4918\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3146\u001b[0m, in \u001b[0;36mEngine._run_ddl_visitor\u001b[0;34m(self, visitorcallable, element, **kwargs)\u001b[0m\n\u001b[1;32m   3145\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run_ddl_visitor\u001b[39m(\u001b[38;5;28mself\u001b[39m, visitorcallable, element, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m-> 3146\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m conn:\n\u001b[1;32m   3147\u001b[0m         conn\u001b[38;5;241m.\u001b[39m_run_ddl_visitor(visitorcallable, element, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3062\u001b[0m, in \u001b[0;36mEngine.begin\u001b[0;34m(self, close_with_result)\u001b[0m\n\u001b[1;32m   3060\u001b[0m     conn \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnect()\n\u001b[1;32m   3061\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 3062\u001b[0m     conn \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43mclose_with_result\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mclose_with_result\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3063\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   3064\u001b[0m     trans \u001b[38;5;241m=\u001b[39m conn\u001b[38;5;241m.\u001b[39mbegin()\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3234\u001b[0m, in \u001b[0;36mEngine.connect\u001b[0;34m(self, close_with_result)\u001b[0m\n\u001b[1;32m   3219\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m, close_with_result\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[1;32m   3220\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return a new :class:`_engine.Connection` object.\u001b[39;00m\n\u001b[1;32m   3221\u001b[0m \n\u001b[1;32m   3222\u001b[0m \u001b[38;5;124;03m    The :class:`_engine.Connection` object is a facade that uses a DBAPI\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3231\u001b[0m \n\u001b[1;32m   3232\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 3234\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_connection_cls\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mclose_with_result\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mclose_with_result\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:96\u001b[0m, in \u001b[0;36mConnection.__init__\u001b[0;34m(self, engine, connection, close_with_result, _branch_from, _execution_options, _dispatch, _has_events, _allow_revalidate)\u001b[0m\n\u001b[1;32m     91\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_has_events \u001b[38;5;241m=\u001b[39m _branch_from\u001b[38;5;241m.\u001b[39m_has_events\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     93\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_dbapi_connection \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m     94\u001b[0m         connection\n\u001b[1;32m     95\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m connection \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m---> 96\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[43mengine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mraw_connection\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     97\u001b[0m     )\n\u001b[1;32m     99\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_transaction \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_nested_transaction \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    100\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__savepoint_seq \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3313\u001b[0m, in \u001b[0;36mEngine.raw_connection\u001b[0;34m(self, _connection)\u001b[0m\n\u001b[1;32m   3291\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mraw_connection\u001b[39m(\u001b[38;5;28mself\u001b[39m, _connection\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m   3292\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return a \"raw\" DBAPI connection from the connection pool.\u001b[39;00m\n\u001b[1;32m   3293\u001b[0m \n\u001b[1;32m   3294\u001b[0m \u001b[38;5;124;03m    The returned object is a proxied version of the DBAPI\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3311\u001b[0m \n\u001b[1;32m   3312\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 3313\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wrap_pool_connect\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_connection\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3283\u001b[0m, in \u001b[0;36mEngine._wrap_pool_connect\u001b[0;34m(self, fn, connection)\u001b[0m\n\u001b[1;32m   3281\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m dialect\u001b[38;5;241m.\u001b[39mdbapi\u001b[38;5;241m.\u001b[39mError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m   3282\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m connection \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 3283\u001b[0m         \u001b[43mConnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_handle_dbapi_exception_noconnection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3284\u001b[0m \u001b[43m            \u001b[49m\u001b[43me\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdialect\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\n\u001b[1;32m   3285\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3286\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   3287\u001b[0m         util\u001b[38;5;241m.\u001b[39mraise_(\n\u001b[1;32m   3288\u001b[0m             sys\u001b[38;5;241m.\u001b[39mexc_info()[\u001b[38;5;241m1\u001b[39m], with_traceback\u001b[38;5;241m=\u001b[39msys\u001b[38;5;241m.\u001b[39mexc_info()[\u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m   3289\u001b[0m         )\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2117\u001b[0m, in \u001b[0;36mConnection._handle_dbapi_exception_noconnection\u001b[0;34m(cls, e, dialect, engine)\u001b[0m\n\u001b[1;32m   2115\u001b[0m     util\u001b[38;5;241m.\u001b[39mraise_(newraise, with_traceback\u001b[38;5;241m=\u001b[39mexc_info[\u001b[38;5;241m2\u001b[39m], from_\u001b[38;5;241m=\u001b[39me)\n\u001b[1;32m   2116\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m should_wrap:\n\u001b[0;32m-> 2117\u001b[0m     \u001b[43mutil\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mraise_\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2118\u001b[0m \u001b[43m        \u001b[49m\u001b[43msqlalchemy_exception\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwith_traceback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexc_info\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfrom_\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43me\u001b[49m\n\u001b[1;32m   2119\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2120\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   2121\u001b[0m     util\u001b[38;5;241m.\u001b[39mraise_(exc_info[\u001b[38;5;241m1\u001b[39m], with_traceback\u001b[38;5;241m=\u001b[39mexc_info[\u001b[38;5;241m2\u001b[39m])\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/compat.py:208\u001b[0m, in \u001b[0;36mraise_\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    205\u001b[0m     exception\u001b[38;5;241m.\u001b[39m__cause__ \u001b[38;5;241m=\u001b[39m replace_context\n\u001b[1;32m    207\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 208\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exception\n\u001b[1;32m    209\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m     \u001b[38;5;66;03m# credit to\u001b[39;00m\n\u001b[1;32m    211\u001b[0m     \u001b[38;5;66;03m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[1;32m    212\u001b[0m     \u001b[38;5;66;03m# as the __traceback__ object creates a cycle\u001b[39;00m\n\u001b[1;32m    213\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m exception, replace_context, from_, with_traceback\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/base.py:3280\u001b[0m, in \u001b[0;36mEngine._wrap_pool_connect\u001b[0;34m(self, fn, connection)\u001b[0m\n\u001b[1;32m   3278\u001b[0m dialect \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdialect\n\u001b[1;32m   3279\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3280\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3281\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m dialect\u001b[38;5;241m.\u001b[39mdbapi\u001b[38;5;241m.\u001b[39mError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m   3282\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m connection \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:310\u001b[0m, in \u001b[0;36mPool.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    302\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    303\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return a DBAPI connection from the pool.\u001b[39;00m\n\u001b[1;32m    304\u001b[0m \n\u001b[1;32m    305\u001b[0m \u001b[38;5;124;03m    The connection is instrumented such that when its\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    308\u001b[0m \n\u001b[1;32m    309\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 310\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_ConnectionFairy\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_checkout\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:868\u001b[0m, in \u001b[0;36m_ConnectionFairy._checkout\u001b[0;34m(cls, pool, threadconns, fairy)\u001b[0m\n\u001b[1;32m    865\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[1;32m    866\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_checkout\u001b[39m(\u001b[38;5;28mcls\u001b[39m, pool, threadconns\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, fairy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    867\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fairy:\n\u001b[0;32m--> 868\u001b[0m         fairy \u001b[38;5;241m=\u001b[39m \u001b[43m_ConnectionRecord\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcheckout\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpool\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    870\u001b[0m         fairy\u001b[38;5;241m.\u001b[39m_pool \u001b[38;5;241m=\u001b[39m pool\n\u001b[1;32m    871\u001b[0m         fairy\u001b[38;5;241m.\u001b[39m_counter \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:476\u001b[0m, in \u001b[0;36m_ConnectionRecord.checkout\u001b[0;34m(cls, pool)\u001b[0m\n\u001b[1;32m    474\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[1;32m    475\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcheckout\u001b[39m(\u001b[38;5;28mcls\u001b[39m, pool):\n\u001b[0;32m--> 476\u001b[0m     rec \u001b[38;5;241m=\u001b[39m \u001b[43mpool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_do_get\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    477\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    478\u001b[0m         dbapi_connection \u001b[38;5;241m=\u001b[39m rec\u001b[38;5;241m.\u001b[39mget_connection()\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/impl.py:256\u001b[0m, in \u001b[0;36mNullPool._do_get\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    255\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_do_get\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 256\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create_connection\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:256\u001b[0m, in \u001b[0;36mPool._create_connection\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    253\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_create_connection\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    254\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Called by subclasses to create a new ConnectionRecord.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 256\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_ConnectionRecord\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:371\u001b[0m, in \u001b[0;36m_ConnectionRecord.__init__\u001b[0;34m(self, pool, connect)\u001b[0m\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__pool \u001b[38;5;241m=\u001b[39m pool\n\u001b[1;32m    370\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m connect:\n\u001b[0;32m--> 371\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m__connect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    372\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfinalize_callback \u001b[38;5;241m=\u001b[39m deque()\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:665\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    663\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfresh \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    664\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 665\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m util\u001b[38;5;241m.\u001b[39msafe_reraise():\n\u001b[1;32m    666\u001b[0m         pool\u001b[38;5;241m.\u001b[39mlogger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError on connect(): \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, e)\n\u001b[1;32m    667\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    668\u001b[0m     \u001b[38;5;66;03m# in SQLAlchemy 1.4 the first_connect event is not used by\u001b[39;00m\n\u001b[1;32m    669\u001b[0m     \u001b[38;5;66;03m# the engine, so this will usually not be set\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py:70\u001b[0m, in \u001b[0;36msafe_reraise.__exit__\u001b[0;34m(self, type_, value, traceback)\u001b[0m\n\u001b[1;32m     68\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exc_info \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m  \u001b[38;5;66;03m# remove potential circular references\u001b[39;00m\n\u001b[1;32m     69\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwarn_only:\n\u001b[0;32m---> 70\u001b[0m         \u001b[43mcompat\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mraise_\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     71\u001b[0m \u001b[43m            \u001b[49m\u001b[43mexc_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     72\u001b[0m \u001b[43m            \u001b[49m\u001b[43mwith_traceback\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexc_tb\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     73\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     75\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m compat\u001b[38;5;241m.\u001b[39mpy3k \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exc_info \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exc_info[\u001b[38;5;241m1\u001b[39m]:\n\u001b[1;32m     76\u001b[0m         \u001b[38;5;66;03m# emulate Py3K's behavior of telling us when an exception\u001b[39;00m\n\u001b[1;32m     77\u001b[0m         \u001b[38;5;66;03m# occurs in an exception handler.\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/util/compat.py:208\u001b[0m, in \u001b[0;36mraise_\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    205\u001b[0m     exception\u001b[38;5;241m.\u001b[39m__cause__ \u001b[38;5;241m=\u001b[39m replace_context\n\u001b[1;32m    207\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 208\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exception\n\u001b[1;32m    209\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m     \u001b[38;5;66;03m# credit to\u001b[39;00m\n\u001b[1;32m    211\u001b[0m     \u001b[38;5;66;03m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[1;32m    212\u001b[0m     \u001b[38;5;66;03m# as the __traceback__ object creates a cycle\u001b[39;00m\n\u001b[1;32m    213\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m exception, replace_context, from_, with_traceback\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/pool/base.py:661\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    659\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    660\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstarttime \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[0;32m--> 661\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdbapi_connection \u001b[38;5;241m=\u001b[39m connection \u001b[38;5;241m=\u001b[39m \u001b[43mpool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_invoke_creator\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    662\u001b[0m     pool\u001b[38;5;241m.\u001b[39mlogger\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCreated new connection \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, connection)\n\u001b[1;32m    663\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfresh \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/create.py:590\u001b[0m, in \u001b[0;36mcreate_engine.<locals>.connect\u001b[0;34m(connection_record)\u001b[0m\n\u001b[1;32m    588\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m connection \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    589\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m connection\n\u001b[0;32m--> 590\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdialect\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcparams\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dev/lib/python3.10/site-packages/sqlalchemy/engine/default.py:597\u001b[0m, in \u001b[0;36mDefaultDialect.connect\u001b[0;34m(self, *cargs, **cparams)\u001b[0m\n\u001b[1;32m    595\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39mcargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mcparams):\n\u001b[1;32m    596\u001b[0m     \u001b[38;5;66;03m# inherits the docstring from interfaces.Dialect.connect\u001b[39;00m\n\u001b[0;32m--> 597\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdbapi\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcparams\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mOperationalError\u001b[0m: (sqlite3.OperationalError) unable to open database file\n(Background on this error at: https://sqlalche.me/e/14/e3q8)"
     ]
    }
   ],
   "source": [
    "from sqlalchemy import Column, Integer, Float, String\n",
    "from sqlalchemy.ext.declarative import declarative_base\n",
    "from sqlalchemy.orm import sessionmaker\n",
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import sqlite3\n",
    "\n",
    "Base = declarative_base()\n",
    "\n",
    "class CAData(Base):\n",
    "    __tablename__ = 'ca_data'\n",
    "    id = Column(Integer, primary_key=True)\n",
    "    station_name = Column(String)\n",
    "    street_address = Column(String)\n",
    "    city = Column(String)\n",
    "    state = Column(String)\n",
    "    zip = Column(Integer)\n",
    "    latitude = Column(Float)\n",
    "    longitude = Column(Float)\n",
    "    access_code = Column(String)\n",
    "    ev_network = Column(String)\n",
    "    ev_connector_types = Column(String)\n",
    "    status_code = Column(String)\n",
    "    ev_pricing = Column(String)\n",
    "    geocode_status = Column(String)\n",
    "\n",
    "class SQLiteManager:\n",
    "    def __init__(self, database_path):\n",
    "        self.database_path = database_path\n",
    "        self.engine = create_engine(f\"sqlite:///{self.database_path}\")\n",
    "        self.session = sessionmaker(bind=self.engine)()\n",
    "\n",
    "    def create_database(self):\n",
    "        Base.metadata.create_all(self.engine)\n",
    "\n",
    "    def import_csv_to_table(self, csv_file_path):\n",
    "        try:\n",
    "            csv_data = pd.read_csv(csv_file_path)\n",
    "            print(\"CSV data read successfully:\")\n",
    "            print(csv_data.head())  \n",
    "            \n",
    "            csv_data.to_sql(\"ca_data\", self.engine, if_exists='replace', index=False)\n",
    "            print(\"Data imported into SQLite table.\")\n",
    "            \n",
    "        except Exception as e:\n",
    "            print(f\"Error importing data to SQLite: {e}\")\n",
    "            raise  # Re-raise the exception to see the full traceback\n",
    "\n",
    "    def execute_query(self, query):\n",
    "        self.session.execute(query)\n",
    "        self.session.commit()\n",
    "\n",
    "    def close_connection(self):\n",
    "        self.session.close()\n",
    "\n",
    "# Usage example\n",
    "csv_file_path = \"data_query/fuel_stations_ca.csv\"\n",
    "database_path = \"data/ca_data.sqlite\"\n",
    "\n",
    "manager = SQLiteManager(database_path)\n",
    "manager.create_database()\n",
    "manager.import_csv_to_table(csv_file_path)\n",
    "\n",
    "# Example query\n",
    "# results = manager.session.query(ReducedData).filter_by(city='YourCity').all()\n",
    "# for result in results:\n",
    "#     print(result.station_name, result.street_address)\n",
    "\n",
    "manager.close_connection()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32bc7760",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
